sync/sync.go

```diff
@@ -246,33 +214,121 @@ func (s *Syncer) sendBlockRequest(peer Peer, id uint32) (chan Block, error) {
 	return ch, s.p.SendAsyncRequest(BLOCK, payload, peer, foo)
 }
 
-func (s *Syncer) sendLayerHashRequest(peer Peer, layer uint32) ([]byte, error) {
-	log.Debug("send Layer hash request Peer: ", peer, " layer: ", layer)
+func (s *Syncer) getLayerBlockIDs(index uint32) (chan uint32, error) {
+
+	m, err := s.getLayerHashes(index)
 
-	data := &pb.LayerHashReq{Layer: layer}
-	payload, err := proto.Marshal(data)
 	if err != nil {
-		log.Error("could not marshall layer hash request")
+		log.Error("could not get LayerHashes for layer: ", index, err)
 		return nil, err
 	}
+	return s.getIdsforHash(m, index)
+}
 
-	msg, err := s.p.SendRequest(LAYER_HASH, payload, peer, s.config.requestTimeout)
-	if err != nil {
-		log.Error("could not send layer hash request ", err)
-		return nil, err
+func (s *Syncer) getIdsforHash(m map[string]Peer, index uint32) (chan uint32, error) {
+	idSet := make(map[uint32]bool, s.config.layerSize)
+	//todo move this to config
+	ch := make(chan []uint32)
+	reqCounter := 0
+
+	for _, v := range m {
+		_, err := s.sendLayerIDsRequest(v, index, ch)
+		if err != nil {
+			log.Error("could not fetch layer ", index, " block ids from peer ", v) // todo recover from this
+			return nil, err
+		} else {
+			reqCounter++
+		}
+	}
+
+	timeout := time.After(s.config.requestTimeout)
+	res := make(chan uint32, s.config.layerSize)
+	defer close(res)
+
+	for {
+		select {
+		case b := <-ch:
+			for _, id := range b {
+				if _, exists := idSet[id]; !exists {
+					idSet[id] = true
+					res <- id
+				}
+			}
+			reqCounter--
+			if reqCounter == 0 {
+				return res, nil
+			}
+		case <-timeout: // Got a timeout! fail with a timeout error
+			if len(res) == 0 {
+				return nil, errors.New("could not get block ids from any peer")
+			}
+			return res, nil
```

Comment:
should this go here? what abour return statement at the end of the function? also, you've timed out here, shouldn't it be an error?